File: DescendantsLibrary1/ControlExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace DescendantsLibrary
{
    public static class ControlExtensions
    {
        public static List<TextBox> TextBoxList(this Control pControl)
        {
            return pControl.Descendants<TextBox>().ToList();
        }
        public static List<CheckBox> CheckBoxList(this Control pControl)
        {
            return pControl.Descendants<CheckBox>().ToList();
        }
        public static List<ComboBox> ComboBoxList(this Control pControl)
        {
            return pControl.Descendants<ComboBox>().ToList();
        }
        public static List<PictureBox> PictureBoxList(this Control pControl)
        {
            return pControl.Descendants<PictureBox>().ToList();
        }
        public static List<Panel> PanelList(this Control pControl)
        {
            return pControl.Descendants<Panel>().ToList();
        }
        public static List<RadioButton> RadioButtonList(this Control pControl)
        {
            return pControl.Descendants<RadioButton>().ToList();
        }
        public static List<NumericUpDown> NumericUpDownList(this Control pControl)
        {
            return pControl.Descendants<NumericUpDown>().ToList();
        }
        /// <summary>
        /// Get a collection of a specific type of control from a control or form.
        /// </summary>
        /// <typeparam name="T">Type of control</typeparam>
        /// <param name="control">Control to traverse</param>
        /// <returns>IEnumerable of T</returns>
        public static IEnumerable<T> Descendants<T>(this Control control) where T : class
        {
            foreach (Control child in control.Controls)
            {
                T thisControl = child as T;
                if (thisControl != null)
                {
                    yield return (T)thisControl;
                }

                if (child.HasChildren)
                {
                    foreach (T descendant in Descendants<T>(child))
                    {
                        yield return descendant;
                    }
                }
            }
        }
    }
}

File: DescendantsLibrary/ControlExtensions.vb
Imports System.Windows.Forms

Public Module ControlExtensions
    ''' <summary>
    ''' Provides access to all controls on a form including
    ''' controls within containers e.g. panel and group-box etc.
    ''' </summary>
    ''' <typeparam name="T"></typeparam>
    ''' <param name="control"></param>
    ''' <returns></returns>
    <Runtime.CompilerServices.Extension>
    Public Iterator Function Descendants(Of T As Class)(control As Control) As IEnumerable(Of T)
        For Each child As Control In control.Controls

            Dim currentChild = TryCast(child, T)
            If currentChild IsNot Nothing Then
                Yield currentChild
            End If

            If child.HasChildren Then
                For Each descendant As T In child.Descendants(Of T)()
                    Yield descendant
                Next
            End If
        Next

    End Function
    <Runtime.CompilerServices.Extension>
    Public Function TextBoxList(pControl As Control) As List(Of TextBox)
        Return pControl.Descendants(Of TextBox)().ToList()
    End Function
    <Runtime.CompilerServices.Extension>
    Public Function CheckBoxList(pControl As Control) As List(Of CheckBox)
        Return pControl.Descendants(Of CheckBox)().ToList()
    End Function
    <Runtime.CompilerServices.Extension>
    Public Function ComboBoxList(pControl As Control) As List(Of ComboBox)
        Return pControl.Descendants(Of ComboBox)().ToList()
    End Function
    <Runtime.CompilerServices.Extension>
    Public Function PictureBoxList(pControl As Control) As List(Of PictureBox)
        Return pControl.Descendants(Of PictureBox)().ToList()
    End Function
    <Runtime.CompilerServices.Extension>
    Public Function PanelList(pControl As Control) As List(Of Panel)
        Return pControl.Descendants(Of Panel)().ToList()
    End Function
    <Runtime.CompilerServices.Extension>
    Public Function RadioButtonList(pControl As Control) As List(Of RadioButton)
        Return pControl.Descendants(Of RadioButton)().ToList()
    End Function
    <Runtime.CompilerServices.Extension>
    Public Function NumericUpDownList(pControl As Control) As List(Of NumericUpDown)
        Return pControl.Descendants(Of NumericUpDown)().ToList()
    End Function
End Module